src/framework/mlt_link.c

```diff
@@ -0,0 +1,129 @@
+/**
+ * \file mlt_link.c
+ * \brief link service class
+ * \see mlt_link_s
+ *
+ * Copyright (C) 2020 Meltytech, LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include "mlt_link.h"
+#include "mlt_frame.h"
+#include "mlt_log.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+/* Forward references to static methods.
+*/
+
+static int producer_get_frame( mlt_producer parent, mlt_frame_ptr frame, int track );
+
+/** Construct a link.
+ *
+ * Sets the resource property to "<link>", the mlt_type to "mlt_producer",
+ * and mlt_service to "link".
+ *
+ * \public \memberof mlt_link_s
+ * \return the new link
+ */
+
+mlt_link mlt_link_init( )
+{
+	mlt_link self = calloc( 1, sizeof( struct mlt_link_s ) );
+	if ( self != NULL )
+	{
+		mlt_producer producer = &self->parent;
+		if ( mlt_producer_init( producer, self ) == 0 )
+		{
+			mlt_properties properties = MLT_PRODUCER_PROPERTIES( producer );
+
+			mlt_properties_set( properties, "resource", "<link>" );
+			mlt_properties_set( properties, "mlt_type", "mlt_producer" );
+			mlt_properties_set( properties, "mlt_service", "link" );
+
+			producer->get_frame = producer_get_frame;
+			producer->close = ( mlt_destructor )mlt_link_close;
+			producer->close_object = self;
+		}
+		else
+		{
+			free( self );
+			self = NULL;
+		}
+	}
+	return self;
+}
+
+/** Connect this link to the next producer.
+ *
+ * \public \memberof mlt_link_s
+ * \param self a link
+ * \param next the producer to get frames from
+ * \param default_profile a profile to use if needed (some links derive their frame rate from the next producer)
+ * \return true on error
+ */
+
+int mlt_link_connect_next( mlt_link self, mlt_producer next, mlt_profile default_profile )
+{
+	self->next = next;
+	if ( self->configure )
+	{
+		self->configure( self, default_profile );
```

Comment:
This is what triggers a link to check the framerate of downstream links to set up its own framerate. For example, a frame doubling deinterlacer would implement configure() and check the framerate of the "next" producer and them double its own framerate.